src/mds/MDCache.cc

```diff
@@ -13050,3 +13096,167 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+      went_wait = true;
+      continue;
+    }
+
+    if (dir->is_ambiguous_auth()) {
+      dir->add_waiter(CDir::WAIT_SINGLEAUTH, new C_MDC_RstatFlush(this, mdr));
+      dout(20) << __func__ << " wait for singleauth: " << *dir << dendl;
+      went_wait = true;
+      continue;
+    }
+
+    if (dir->is_auth()) {
+      dout(20) << "dir " << *dir << "is auth, skipping" << dendl;
+      dirs.clear();
+      get_wouldbe_subtree_bounds(dir, dirs);
+      list<CDir*> tmpls(dirs.begin(), dirs.end());
+      drls.splice(drls.end(), tmpls, tmpls.begin(), tmpls.end());
+      continue;
+    }
+
+    if (mdr->more()->queried_for_rstat_propagation.count(dir)) {
+      auto last_auth = mdr->more()->queried_for_rstat_propagation[dir];
+      if (last_auth == dir->get_dir_auth().first) {
+	dout(20) << "dir " << *dir << "has been queried, skipping" << dendl;
+	continue;
+      } else {
+	// the subtree must have been migrated to another mds other
+	// than the one on which we find it on in the last run
+	mdr->more()->queried_for_rstat_propagation.erase(dir);
+	mdr->more()->mds_queried_dirs_map[last_auth].erase(dir);
+	if (!mdr->more()->mds_queried_dirs_map[last_auth].size()) {
+	  mdr->more()->waiting_on_slave.erase(last_auth);
+	  mdr->more()->mds_queried_dirs_map.erase(last_auth);
+	}
+      }
+    }
+
+    if (!mds->mdsmap->is_up(dir->get_dir_auth().first)) {
+      mds->wait_for_active_peer(dir->get_dir_auth().first, new C_MDS_RetryRequest(this, mdr));
+      went_wait = true;
+      continue;
+    }
+
+    dout(20) << __func__ << " found replica for: " << *dir << ", quering" << dendl;
+
+    // Since there can be at most one slave request with a specific reqid(assert
+    // "ceph_assert(mdr->slave_request == 0)" in Server::handle_slave_request),
+    // and we may have to send multiple slave requests to the same mds(if there are
+    // multiple subtrees whose auth are the same mds), we have to start an internal
+    // request for each slave request, so that slave requests have different reqid.
+    //
+    // FIXME this is a temporary solution. Once mds-mds versioned message is available,
+    // we can fix this by adding a new parent_reqid field which holds the reqid of the
+    // request that created the slave request, which makes the current reqid field only serves
+    // as a unique slave request id. With this approach, we can satisfy the assert
+    // "ceph_assert(mdr->slave_request == 0)" in Server::handle_slave_request() method.
+    MDRequestRef internal_mdr = request_start_internal(CEPH_MDS_OP_SLAVERSTATPROPAGATE);
+    internal_mdr->parent_mdr = mdr;
+    auto req = make_message<MMDSSlaveRequest>(internal_mdr->reqid, internal_mdr->attempt,
+        MMDSSlaveRequest::OP_PROPAGATERSTATS);
+    dir->get_inode()->make_path(req->srcdnpath);
+    req->op_stamp = mdr->get_op_stamp();
+    mds->send_message_mds(req, dir->get_dir_auth().first);
+    internal_mdr->more()->queried_dir = dir;
```

Comment:
why not just set origin base of rstat flush. If we do this, there is no reason to create multiple slave requests.

Reply:
@ukernel I'll try to follow your way:-)